### PR TITLE
Add matrix_rank Op and its GPU and CPU kernel

### PR DESCRIPTION
**Pull request by 0x45f**
### PR types

New features
### PR changes

OPs
### Describe

Add paddle.linalg.matrix_rank() function, matrix_rank Op and it's GPU/CPU kernel.

### CODE REVIEW

File: paddle/fluid/operators/matrix_rank_op.cc [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+
+#include "paddle/fluid/framework/ddim.h"
+#include "paddle/fluid/operators/elementwise/elementwise_op_function.h"
+#include "paddle/fluid/operators/elementwise/svd_helper.h"
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cc]: svd_helper.h 不应该放到 elementwise 目录下吧？

File: paddle/fluid/operators/matrix_rank_op.cc [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+#ifdef PADDLE_WITH_MKLDNN
+#include "paddle/fluid/platform/mkldnn_helper.h"
+#endif
+
+namespace paddle {
+namespace operators {
+using DDim = framework::DDim;
+
+DDim InputBatchDim(const DDim& dim_x) {
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cc]: 若这些方法只在此文件里用到，建议使用匿名namespace 包裹一下：
```cpp
namespace{

XXXX
}
```

另外函数名字一般是 动宾形式，如`GetInputBatchDim`

File: paddle/fluid/operators/matrix_rank_op.cc [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+  auto x_vec = framework::vectorize(dim_x);
+  if (x_vec.size() == 2) {
+    return framework::make_ddim({1});
+  }
+  x_vec.erase(x_vec.end() - 2, x_vec.end());
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cc:35]: 如果 x_vec 的size可能会小于1么？这里是否需要加一下条件判断呢？

File: paddle/fluid/operators/matrix_rank_op.cc [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+  return framework::make_ddim(x_vec);
+}
+
+DDim EigenvalueDim(const DDim& dim, int k) {
+  auto vec = framework::vectorize(dim);
+  vec.erase(vec.end() - 2, vec.end());
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cc]: 同上

File: paddle/fluid/operators/matrix_rank_op.cc [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+DDim NewAxisDim(const DDim& dim, int k) {
+  auto vec = framework::vectorize(dim);
+  vec.push_back(k);
+  return framework::make_ddim(vec);
+}
+
+DDim RemoveLastDim(const DDim& dim) {
+  auto vec = framework::vectorize(dim);
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cc]: dim 是否可能为空？

File: paddle/fluid/operators/matrix_rank_op.cc [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+  if (vec.size() == 1) {
+    return framework::make_ddim({1});
+  }
+  vec.erase(vec.end() - 1, vec.end());
+  return framework::make_ddim(vec);
+}
+
+DDim UDDim(const DDim& x_dim, int k) {
+  auto x_vec = framework::vectorize(x_dim);
+  x_vec[x_vec.size() - 1] = k;
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cc]: x_vec.size() 若为0呢？

File: paddle/fluid/operators/matrix_rank_op.cc [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    OP_INOUT_CHECK(ctx->HasInput("X"), "Input", "X", "MatrixRank");
+    OP_INOUT_CHECK(ctx->HasOutput("Out"), "Output", "Out", "MatrixRank");
+    auto dim_x = ctx->GetInputDim("X");
+    PADDLE_ENFORCE_GE(dim_x.size(), 2,
+                      platform::errors::InvalidArgument(
+                          "the dims of input must greater than 2"));
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cc]: -> must be，注意语法， the -> The

File: paddle/fluid/operators/matrix_rank_op.cc [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+    This operator is used to perform MatrixRank operation for batched matrics.
+    $$out = matrix_rank(X, tol, hermitian)$$
+    )DOC");
+  }
+};
+
+template <typename T>
+void BatchEigenvalues(const T* x_data, T* eigenvalues_data, int batches,
+                      int rows, int cols, int k) {
+  T* input = const_cast<T*>(x_data);
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cc:125]: 这里添加一下为什么需要const_cast 的原因说明

File: paddle/fluid/operators/matrix_rank_op.cc [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+    for (int j = 0; j < k; j++) {
+      *(eigenvalues_data + i * k + j) = eigenvalues[j];
+    }
+  }
+}
+
+template <typename T>
+void BatchSVD(const T* x_data, T* eigenvalues_data, int batches, int rows,
+              int cols, int k) {
+  T* input = const_cast<T*>(x_data);
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cc:145]: 同上

File: paddle/fluid/operators/matrix_rank_op.cc (hunk 6 above)
Comment (Aurelius84) [reply paddle/fluid/operators/matrix_rank_op.cc]: 统一回复：若在调用上述函数前做过条件检查，请忽略

File: paddle/fluid/operators/matrix_rank_op.cc [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+      ElementwiseComputeEx<CompareFunctor<T>, platform::CPUDeviceContext, T,
+                           int>(context, &eigenvalue_tensor, &tol_tensor, axis,
+                                CompareFunctor<T>(), &compare_result);
+    } else {
+      ElementwiseComputeEx<InverseCompareFunctor<T>, platform::CPUDeviceContext,
+                           T, int>(context, &eigenvalue_tensor, &tol_tensor,
+                                   axis, InverseCompareFunctor<T>(),
+                                   &compare_result);
+    }
+    auto dito_int =
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cc:239]: 这里貌似不需要再定义一个dito_int，可以复用 dito_T， T模板参数只有在pow中用到了

File: paddle/fluid/operators/matrix_rank_op.cu [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#ifndef PADDLE_WITH_HIP
+// HIP not support cusolver
+#include <thrust/device_vector.h>
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cu]: 这个thrust/device_vector.h 用到了么？

File: paddle/fluid/operators/matrix_rank_op.cu [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+#include <algorithm>
+#include <vector>
+#include "paddle/fluid/memory/memory.h"
+#include "paddle/fluid/operators/cholesky_op.h"
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cu]: cholesky_op.h也没有用到吧，建议check下哪些头文件是不需要的

File: paddle/fluid/operators/matrix_rank_op.cu [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+namespace paddle {
+namespace operators {
+
+template <typename T>
+class MatrixRankGPUKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto& dev_ctx =
+        context.template device_context<platform::CUDADeviceContext>();
+    // get input/output
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cu]: 这些注释可以移除掉，从代码中就可以看出含义，不需要额外的注释，下面也是。

File: paddle/fluid/operators/matrix_rank_op.cu [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+    const Tensor* x = context.Input<Tensor>("X");
+    auto* x_data = x->data<T>();
+    auto* out = context.Output<Tensor>("Out");
+    out->mutable_data<int64_t>(context.GetPlace());
+    bool hermitian = context.Attr<bool>("hermitian");
+    // get shape
+    auto dim_x = x->dims();
+    auto dim_out = out->dims();
+    // auto dim_atol_tensor = atol_tensor->dims();
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cu]: 无用的代码要清理掉

File: paddle/fluid/operators/matrix_rank_op.cu [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+      ElementwiseComputeEx<CompareFunctor<T>, platform::CUDADeviceContext, T,
+                           int64_t>(context, &eigenvalue_tensor, &tol_tensor,
+                                    axis, CompareFunctor<T>(), &compare_result);
+    } else {
+      ElementwiseComputeEx<InverseCompareFunctor<T>,
+                           platform::CUDADeviceContext, T, int64_t>(
+          context, &eigenvalue_tensor, &tol_tensor, axis,
+          InverseCompareFunctor<T>(), &compare_result);
+    }
+    auto dito_int =
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.cu:143]: 同上，这里是不是也可以复用dito_T?

File: paddle/fluid/operators/matrix_rank_op.h [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+DDim EigenvalueDim(const DDim& dim, int k);
+
+DDim NewAxisDim(const DDim& dim, int k);
+
+DDim RemoveLastDim(const DDim& dim);
+
+DDim UDDim(const DDim& x_dim, int k);
+
+DDim VHDDim(const DDim& x_dim, int k);
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.h]: 上面这些接口只在.cc用到了，直接定义在.cc即可。

File: paddle/fluid/operators/matrix_rank_op.h [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+template <typename T>
+struct CompareFunctor {
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.h]: CompareFunctor -> GreaterThanFunctor

File: paddle/fluid/operators/matrix_rank_op.h [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+  HOSTDEVICE int operator()(const T& a, const T& b) const { return a > b; }
+};
+
+template <typename T>
+struct InverseCompareFunctor {
```
Comment (Aurelius84) [inline paddle/fluid/operators/matrix_rank_op.h]: InverseCompareFunctor -> LessThanFunctor

File: python/paddle/fluid/tests/unittests/test_matrix_rank_op.py [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+    def init_data(self):
+        self.x = np.eye(3, dtype=np.float32)
+        self.tolTensor = None
+        self.tol = 0.1
+        self.use_default_tol = False
+        self.hermitian = True
+        self.out = np.linalg.matrix_rank(self.x, self.tol, self.hermitian)
+
+
+class TestMatrixRankOP1(TestMatrixRankOP):
```
Comment (Aurelius84) [inline python/paddle/fluid/tests/unittests/test_matrix_rank_op.py:58]: 新增一些其他corner case，如shape为 [N], [N, 1]、[1, N]

File: python/paddle/tensor/linalg.py [hunk 20]
```diff
@@ -785,6 +785,93 @@ def cholesky(x, upper=False, name=None):
...
 
 
+def matrix_rank(x, tol=None, hermitian=False, name=None):
+    r"""
+    Computes the numerical rank of a matrix.
+
+    The matrix rank is computed as the number of singular values (or eigenvalues in absolute value when hermitian= True) that 
+    are greater than the specified tol threshold.
+
+    Supports input of float, double dtypes. Also supports batches of matrices, and if x is a batch of matrices then the output
```
Comment (Aurelius84) [inline python/paddle/tensor/linalg.py]: 注意英文语法，Also supports -> It also supports.

File: python/paddle/tensor/linalg.py [hunk 21]
```diff
@@ -785,6 +785,93 @@ def cholesky(x, upper=False, name=None):
...
+                                      'use_default_tol', True)
+        elif isinstance(tol, Variable):
+            if tol.dtype != x.dtype:
+                return _C_ops.matrix_rank(x,
+                                          cast(tol, x.dtype), 'hermitian',
+                                          hermitian, 'use_default_tol', False)
+            return _C_ops.matrix_rank(x, tol, 'hermitian', hermitian,
+                                      'use_default_tol', False)
+        else:
+            return _C_ops.matrix_rank(x, None, "tol", tol, 'hermitian',
```
Comment (Aurelius84) [inline python/paddle/tensor/linalg.py]: 上面这段if else 逻辑建议修改为最后调用 _C_ops，如：
```python

if tol is None:
    use_default_tol = True
    XXX = YYY

return _C_ops.matrix_rank(XXX)
```

File: python/paddle/tensor/linalg.py [hunk 22]
```diff
@@ -785,6 +785,97 @@ def cholesky(x, upper=False, name=None):
...
+
+    If tol is not specified and x is a matrix of dimensions (m, n), the tolerance is set to be tol=sigma * max(m,n) * eps where 
+    sigma is the largest singular value (or eigenvalue in absolute value when hermitian= True), and eps is the epsilon value for the 
+    dtype of x. If x is a batch of matrices, tol is computed this way for every element of the batch.
+
+    Args:
+        x (Tensor): tensor of shape [*, m, n] where * is zero or more batch dimensions.
+        tol (float,Tensor,optional): the tolerance value. See above for the value it takes when None. Default: None.
+        hermitian (bool,optional): indicates whether x is Hermitian. Default: False.
+        name (str, optional): Name for the operation (optional, default is None). For more information, please refer to :ref:`api_guide_Name`.
```
Comment (Aurelius84) [inline python/paddle/tensor/linalg.py:806]: 这里 api_guide_Name` 与 . 之间要加个空格，不然预览会有问题

**Review by Aurelius84 — DISMISSED** [review]
LGTM

**Review by zhiqiu — DISMISSED** [review]
LGTM for op_function_generator.cc and ShareDataWith

**Review by XiaoguangHu01 — DISMISSED** [review]
LG API

**Review by dingjiaweiww — DISMISSED** [review]

**Review by qingqing01 — COMMENTED** [review]
LGTM for API. but need to update doc commets

**Review by qingqing01 — APPROVED** [review]
LGTM for API

**Review by Aurelius84 — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by zhhsplendid — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]

**Review by dingjiaweiww — APPROVED** [review]
